### PR TITLE
Add TextTextCLIP

### PR DESCRIPTION
**Pull request by lingjzhu**
This pull request adds TextTextCLIP (CLIP-like text-to-text contrastive retrieval model) to the main branch. 
It is still a work in progress. 

Tasks
- [X] Add a config file for TextTextCLIP
- [X] Add TextTextCLIP in `model.py`
- [X] Modify `factory.py` to load model
- [X] Modify  `data.py` to load text data
- [ ] Modify `main.py' to  train TextTextCLIP
- [X] Test loading TextTextCLIP 
- [ ] Test loading text-pair data. 
- [ ] Test dummy training

### CODE REVIEW

File: src/training/params.py [hunk 1]
```diff
@@ -326,13 +326,54 @@ def parse_args(args):
         action='store_true',
         help="Freeze BatchNorm running stats in image tower for any locked layers.",
     )
+    parser.add_argument(
+        "--lock-doc",
```
Comment (rom1504) [inline src/training/params.py:330]: what about a "lock-tower-1" and "lock-tower-2" param instead ?
query/doc is not the only thing

Comment (lingjzhu) [reply src/training/params.py:342]: Sure. I will rename these variables immediately. Do you think we should also rename variables like `image_features` in `train.py` and `main.py'? Will renaming variables cause confusion for the original text-image model? Currently `train.py` is fully compatible with TextTextCLIP. No modification other than renaming variables is needed.

Comment (lingjzhu) [reply src/training/params.py:342]: I have tested TextTextCLIP locally and it seems to work smoothly. But the script fails some tests here. I will make adjustments.

File: src/training/params.py [hunk 2]
```diff
@@ -326,13 +338,54 @@ def parse_args(args):
         action='store_true',
         help="Freeze BatchNorm running stats in image tower for any locked layers.",
     )
+    parser.add_argument(
```
Comment (rom1504) [inline src/training/params.py:341]: wonder if this can be reconciled with the normal locking params

File: src/training/data.py [hunk 3]
```diff
@@ -27,6 +27,54 @@
     hvd = None
 
 
+class TextPairDataset(Dataset):
+    def __init__(self, input_filename, text_a_key, text_b_key, tokenizer=None):
+        logging.debug(f'Loading parquet data from {input_filename}.')
+        df = pd.read_parquet(input_filename)
```
Comment (rom1504) [inline src/training/data.py:33]: this is unlikely to scale

File: src/open_clip/model.py [hunk 4]
```diff
@@ -248,6 +248,47 @@ def forward(self, image, text):
         return image_features, text_features, self.logit_scale.exp()
 
 
+
+class TextTextCLIP(nn.Module):
```
Comment (rom1504) [inline src/open_clip/model.py:252]: this is completely duplicated from above class, I wonder if we could reconcile it

File: src/open_clip/factory.py [hunk 5]
```diff
@@ -263,18 +285,20 @@ def create_model_from_pretrained(
...
 
-    image_mean = image_mean or getattr(model.visual, 'image_mean', None)
-    image_std = image_std or getattr(model.visual, 'image_std', None)
-    preprocess = image_transform(
-        model.visual.image_size,
-        is_train=False,
-        mean=image_mean,
-        std=image_std
-    )
+    if not text_to_text:
```
Comment (rom1504) [inline src/open_clip/factory.py:294]: what about checking if model.visual exists instead ?

File: src/open_clip/factory.py [hunk 6]
```diff
@@ -179,9 +193,10 @@ def create_model(
         if precision in ("fp16", "bf16"):
             convert_weights_to_lp(model, dtype=torch.bfloat16 if precision == 'bf16' else torch.float16)
 
+        if not text_to_text:
```
Comment (rom1504) [inline src/open_clip/factory.py:196]: what about checking if model.visual exists instead ?

File: src/open_clip/model.py (hunk 4 above)
Comment (lingjzhu) [reply src/open_clip/model.py:252]: Maybe we could use this one as the general model? It's not specific about modality. We can refer to `image_features` as `features_a` and `text_features` as `features_b`.

File: src/open_clip/factory.py (hunk 5 above)
Comment (lingjzhu) [reply src/open_clip/factory.py:294]: If we merge `TextTextCLIP` to `CustomCLIP` to get a more general model, then `model.visual` might not exist at all?

File: src/training/data.py (hunk 3 above)
Comment (lingjzhu) [reply src/training/data.py:33]: Right. I will update this. Do you have any suggestions?

**Comment by rom1504** [comment]
Can you rebase on master ?

**Comment by lingjzhu** [comment]
> Can you rebase on master ?

Done! I've been setting up training in the computing cluster and fixing typos here and there. The code is mostly OK now. It should be ready once training has been successfully completed. I will post `wandb` logs for model training soon.

**Comment by lingjzhu** [comment]
@rom1504 Here is the latest code for TextTextCLIP. It is tested on the stability cluster and the evaluation code is also included. I have added an example script at `docs/script_examples/text_example.sh`. Let me know if it is OK to merge or more revisions are needed. Thank you!

**Comment by rom1504** [comment]
This would really benefits from reviews @rwightman @mitchellnw @iejMac 
It was trained quite decently
Getting it in a good enough state and merged would go a long way to expend the applicability space of openclip

**Comment by rwightman** [comment]
@rom1504 k, will try and look at it soon

File: src/open_clip/loss.py [hunk 7]
```diff
@@ -99,33 +99,44 @@ def get_ground_truth(self, device, num_logits) -> torch.Tensor:
...
+            logits_per_feature_b = logit_scale * features_b @ features_a.T
+
+       
+        return logits_per_feature_a, logits_per_feature_b
+
+
+    def forward(self, image_features=None, text_features=None, logit_scale=None, text_a_features=None, text_b_features=None, output_dict=False):
         
-        return logits_per_image, logits_per_text
+        if image_features is not None and text_features is not None:
```
Comment (daniel-z-kaplan) [inline src/open_clip/loss.py:125]: This section needs a check for the case of both image/text are none, since that's something new that can occur now I think?

File: src/open_clip/loss.py [hunk 8]
```diff
@@ -99,33 +99,44 @@ def get_ground_truth(self, device, num_logits) -> torch.Tensor:
...
-            logits_per_image = logit_scale * image_features @ text_features.T
-            logits_per_text = logit_scale * text_features @ image_features.T
+            logits_per_feature_a = logit_scale * features_a @ features_b.T
+            logits_per_feature_b = logit_scale * features_b @ features_a.T
+
+       
+        return logits_per_feature_a, logits_per_feature_b
+
+
+    def forward(self, image_features=None, text_features=None, logit_scale=None, text_a_features=None, text_b_features=None, output_dict=False):
```
Comment (daniel-z-kaplan) [inline src/open_clip/loss.py:124]: This section feels a little awkward, maybe it should just always take features_a and features_b, not optional, like the prior code?
Same with logit_scale, why is that optional now?

**Comment by rwightman** [comment]
overall things look pretty good, I'm trying to get over a mental block re the loss naming, I reliaze why the feature_a/b changes were made to the loss but I feel it harms ease of understanding re the  most common use case, especially for newcomers to the core loss fn... hmm hnm

File: src/open_clip/loss.py (hunk 8 above)
Comment (lingjzhu) [reply src/open_clip/loss.py:124]: Thank you for your comments! This is an artifact of trying to reconcile the original CLIP and the text model. But I agree with you that this is a bit awkward. I could add a loss wrapper for the text model. What do you think?

By the way, do you have any ideas as to the naming of the text model? TextTextCLIP sounds a bit redundant to me. How about CTTP? CLLP? LATTE (contrastive LAnguage-To-TExt pretraining)? CTP (contrastive text pretraining)? Do you have a better suggestions?
